[InstCombine] limit icmp sgt (shl nsw X, C1), C0 --> icmp sgt X, C0 >> C1 to hasOneUse

ChunyuLiao opened the pull request:
Fix: https://github.com/llvm/llvm-project/issues/30121

Testcase ir:  https://gcc.godbolt.org/z/oKTPE7v48
```
  %cmp11 = icmp sgt i32 %n, 0
  br i1 %cmp11, label %for.body.preheader, label %for.cond.cleanup

for.body.preheader:                               ; preds = %entry
  %mul = shl i32 %n, 4
  %smax = tail call i32 @llvm.smax.i32(i32 %mul, i32 1)
  %wide.trip.count = zext nneg i32 %smax to i64
  br label %for.body
```

If add the limitations, @llvm.smax shouldn't be needed.

```
  %mul = shl nsw i32 %n, 4
  %cmp11 = icmp sgt i32 %mul, 0
  br i1 %cmp11, label %for.body.preheader, label %for.cond.cleanup

for.body.preheader:                               ; preds = %entry
  %smax = tail call i32 @llvm.smax.i32(i32 %mul, i32 1) --- this shouldn't be needed

```
If we need icmp sgt (shl nsw X, C1), C0 --> icmp sgt X, C0 >> C1， perhaps we can implement this at DAG combine.

ChunyuLiao commented:
> I don't think this is the right way to fix this issue. If I'm reading the transformation log right, the problem seems to be that we lose nsw information in IndVarSimplify?

Could you help give more information? Thanks.
My understanding is that shl's nsw should be retained in IndVarSimplify.

Violent do try. Keep shl's nsw, but it doesn't fix the issue.
```
diff --git a/llvm/lib/Analysis/ValueTracking.cpp b/llvm/lib/Analysis/ValueTracking.cpp
index 8c29c242215d..b43feb621ad6 100644
--- a/llvm/lib/Analysis/ValueTracking.cpp
+++ b/llvm/lib/Analysis/ValueTracking.cpp
@@ -7309,7 +7309,7 @@ static bool programUndefinedIfUndefOrPoison(const Value *V,
 
       // If an operand is poison and propagates it, mark I as yielding poison.
       for (const Use &Op : I.operands()) {
-        if (YieldsPoison.count(Op) && propagatesPoison(Op)) {
+        if (propagatesPoison(Op)) {
           YieldsPoison.insert(&I);
           break;
         }
```

ChunyuLiao commented:
I guess I found where SCEV takes effect against this patch in:
https://github.com/llvm/llvm-project/blob/main/llvm/lib/Analysis/ScalarEvolution.cpp#L11872
```
std::optional<APInt>
ScalarEvolution::computeConstantDifference(const SCEV *More, const SCEV *Less) {
  // We avoid subtracting expressions here because this function is usually
  // fairly deep in the call stack (i.e. is called many times).

  // X - X = 0.
  if (More == Less)
    return APInt(getTypeSizeInBits(More->getType()), 0);
```
There are some optimizations for SCEVAddRecExpr and SCEVAddExpr in computeConstantDifference, but in this example it's: SCEVType = llvm::scUnknown, so it can't be handled

goldsteinn commented on llvm/test/Transforms/IndVarSimplify/pr30121.ll:
```diff
@@ -0,0 +1,53 @@
+  %inc = add nuw nsw i32 %i.012, 1
+  %cmp = icmp slt i32 %inc, %mul
+  br i1 %cmp, label %for.body, label %for.cond.cleanup.loopexit
+}
```
can you pre-commit the tests?

ChunyuLiao replied:
pre test pr: https://github.com/llvm/llvm-project/pull/83770

goldsteinn replied:
Err Im sorry, I was unclear. I mean can you just split it into two commits?
I.e 
commit 1: add new tests
commit 2: impl + update tests.

An example might be: https://github.com/llvm/llvm-project/pull/83739/commits

ChunyuLiao replied:
thanks

ChunyuLiao committed:
[IndVarSimplify]Add pre-commit tests for PR74318. NFC.

ChunyuLiao committed:
[InstCombine] limit icmp sgt (shl nsw X, C1), C0 --> icmp sgt X, C0 >> C1 to hasOneUse

Fix: https://github.com/llvm/llvm-project/issues/30121

Testcase ir:  https://gcc.godbolt.org/z/oKTPE7v48
```
   %cmp11 = icmp sgt i32 %n, 0
   br i1 %cmp11, label %for.body.preheader, label %for.cond.cleanup

   for.body.preheader:                               ; preds = %entry
   %mul = shl i32 %n, 4
   %smax = tail call i32 @llvm.smax.i32(i32 %mul, i32 1)
   %wide.trip.count = zext nneg i32 %smax to i64
   br label %for.body
```

If add the limitations, @llvm.smax shouldn't be needed.

```
  %mul = shl nsw i32 %n, 4
  %cmp11 = icmp sgt i32 %mul, 0
  br i1 %cmp11, label %for.body.preheader, label %for.cond.cleanup

  for.body.preheader:                               ; preds = %entry
  %smax = tail call i32 @llvm.smax.i32(i32 %mul, i32 1) --- this shouldn't be needed

```
  If we need icmp sgt (shl nsw X, C1), C0 --> icmp sgt X, C0 >> C1，
  perhaps we can implement this at DAG combine.
